File: oauth2/typescript/src/core/state.ts
/**
 * State Manager
 *
 * CSRF protection via state parameter management.
 */

import * as crypto from "crypto";
import { AuthorizationError } from "../error";
import { StateMetadata } from "../types";

/**
 * State manager interface (for dependency injection).
 */
export interface StateManager {
  /**
   * Generate a new state with metadata.
   */
  generate(metadata: Omit<StateMetadata, "createdAt" | "expiresAt">): string;

  /**
   * Validate and consume a state.
   */
  consume(state: string): StateMetadata | null;

  /**
   * Check if state exists (without consuming).
   */
  exists(state: string): boolean;

  /**
   * Clean up expired states.
   */
  cleanup(): void;
}

/**
 * In-memory state manager implementation.
 */
export class InMemoryStateManager implements StateManager {
  private states: Map<string, StateMetadata> = new Map();
  private stateExpiration: number;
  private entropyBytes: number;

  constructor(options?: { expirationSeconds?: number; entropyBytes?: number }) {
    this.stateExpiration = (options?.expirationSeconds ?? 600) * 1000;
    this.entropyBytes = options?.entropyBytes ?? 16; // 128 bits minimum
  }

  generate(metadata: Omit<StateMetadata, "createdAt" | "expiresAt">): string {
    const state = this.generateSecureState();
    const now = new Date();

    const fullMetadata: StateMetadata = {
      ...metadata,
      createdAt: now,
      expiresAt: new Date(now.getTime() + this.stateExpiration),
    };

    this.states.set(state, fullMetadata);
    return state;
  }

  consume(state: string): StateMetadata | null {
    const metadata = this.states.get(state);
    if (!metadata) {
      return null;
    }

    // Remove state (one-time use)
    this.states.delete(state);

    // Check expiration
    if (new Date() > metadata.expiresAt) {
      return null;
    }

    return metadata;
  }

  exists(state: string): boolean {
    const metadata = this.states.get(state);
    if (!metadata) {
      return false;
    }
    return new Date() <= metadata.expiresAt;
  }

  cleanup(): void {
    const now = new Date();
    for (const [state, metadata] of this.states) {
      if (now > metadata.expiresAt) {
        this.states.delete(state);
      }
    }
  }

  private generateSecureState(): string {
    const bytes = crypto.randomBytes(this.entropyBytes);
    return bytes.toString("base64url");
  }
}

/**
 * Mock state manager for testing.
 */
export class MockStateManager implements StateManager {
  private states: Map<string, StateMetadata> = new Map();
  private generateHistory: StateMetadata[] = [];
  private consumeHistory: string[] = [];
  private nextState?: string;

  /**
   * Set the next state to generate.
   */
  setNextState(state: string): this {
    this.nextState = state;
    return this;
  }

  /**
   * Pre-populate a state.
   */
  setState(state: string, metadata: StateMetadata): this {
    this.states.set(state, metadata);
    return this;
  }

  /**
   * Get generate history.
   */
  getGenerateHistory(): StateMetadata[] {
    return [...this.generateHistory];
  }

  /**
   * Get consume history.
   */
  getConsumeHistory(): string[] {
    return [...this.consumeHistory];
  }

  generate(metadata: Omit<StateMetadata, "createdAt" | "expiresAt">): string {
    const state = this.nextState ?? `mock-state-${Date.now()}`;
    this.nextState = undefined;

    const now = new Date();
    const fullMetadata: StateMetadata = {
      ...metadata,
      createdAt: now,
      expiresAt: new Date(now.getTime() + 600000),
    };

    this.states.set(state, fullMetadata);
    this.generateHistory.push(fullMetadata);
    return state;
  }

  consume(state: string): StateMetadata | null {
    this.consumeHistory.push(state);
    const metadata = this.states.get(state);
    if (metadata) {
      this.states.delete(state);
    }
    return metadata ?? null;
  }

  exists(state: string): boolean {
    return this.states.has(state);
  }

  cleanup(): void {
    // No-op for mock
  }
}

/**
 * Validate state using constant-time comparison.
 */
export function validateState(
  received: string,
  expected: string
): boolean {
  if (received.length !== expected.length) {
    // Still do comparison to avoid timing attack
    crypto.timingSafeEqual(
      Buffer.from(received),
      Buffer.from(received)
    );
    return false;
  }

  return crypto.timingSafeEqual(
    Buffer.from(received),
    Buffer.from(expected)
  );
}

/**
 * Create production state manager.
 */
export function createStateManager(
  expirationSeconds?: number
): StateManager {
  return new InMemoryStateManager({ expirationSeconds });
}

/**
 * Create mock state manager for testing.
 */
export function createMockStateManager(): MockStateManager {
  return new MockStateManager();
}

File: oauth2/rust/src/core/state.rs
//! State Management
//!
//! OAuth2 state parameter generation and validation.

use rand::Rng;
use std::collections::HashMap;
use std::sync::Mutex;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use crate::error::{AuthorizationError, OAuth2Error};
use crate::types::StateMetadata;

/// State manager interface (for dependency injection).
pub trait StateManager: Send + Sync {
    /// Generate a new state parameter with metadata.
    fn generate(&self, metadata: StateMetadata) -> String;

    /// Validate and consume a state parameter.
    fn consume(&self, state: &str) -> Option<StateMetadata>;

    /// Check if state exists.
    fn exists(&self, state: &str) -> bool;

    /// Clear expired states.
    fn clear_expired(&self);
}

/// In-memory state manager implementation.
pub struct InMemoryStateManager {
    states: Mutex<HashMap<String, StateMetadata>>,
    max_age_ms: u64,
}

impl InMemoryStateManager {
    /// Create new state manager with default TTL (10 minutes).
    pub fn new() -> Self {
        Self::with_max_age(Duration::from_secs(600))
    }

    /// Create state manager with custom TTL.
    pub fn with_max_age(max_age: Duration) -> Self {
        Self {
            states: Mutex::new(HashMap::new()),
            max_age_ms: max_age.as_millis() as u64,
        }
    }

    fn generate_random_state() -> String {
        let mut rng = rand::thread_rng();
        let bytes: [u8; 32] = rng.gen();
        base64::Engine::encode(&base64::engine::general_purpose::URL_SAFE_NO_PAD, bytes)
    }
}

impl Default for InMemoryStateManager {
    fn default() -> Self {
        Self::new()
    }
}

impl StateManager for InMemoryStateManager {
    fn generate(&self, metadata: StateMetadata) -> String {
        let state = Self::generate_random_state();
        self.states
            .lock()
            .unwrap()
            .insert(state.clone(), metadata);
        state
    }

    fn consume(&self, state: &str) -> Option<StateMetadata> {
        let mut states = self.states.lock().unwrap();
        let metadata = states.remove(state)?;

        // Check if expired
        if metadata.is_expired(self.max_age_ms) {
            return None;
        }

        Some(metadata)
    }

    fn exists(&self, state: &str) -> bool {
        let states = self.states.lock().unwrap();
        if let Some(metadata) = states.get(state) {
            !metadata.is_expired(self.max_age_ms)
        } else {
            false
        }
    }

    fn clear_expired(&self) {
        let mut states = self.states.lock().unwrap();
        states.retain(|_, metadata| !metadata.is_expired(self.max_age_ms));
    }
}

/// Mock state manager for testing.
#[derive(Default)]
pub struct MockStateManager {
    states: Mutex<HashMap<String, StateMetadata>>,
    generate_history: Mutex<Vec<StateMetadata>>,
    consume_history: Mutex<Vec<String>>,
    next_state: Mutex<Option<String>>,
}

impl MockStateManager {
    /// Create new mock state manager.
    pub fn new() -> Self {
        Self::default()
    }

    /// Set the next state to generate.
    pub fn set_next_state(&self, state: String) -> &Self {
        *self.next_state.lock().unwrap() = Some(state);
        self
    }

    /// Pre-populate a state.
    pub fn add_state(&self, state: String, metadata: StateMetadata) -> &Self {
        self.states.lock().unwrap().insert(state, metadata);
        self
    }

    /// Get generate history.
    pub fn get_generate_history(&self) -> Vec<StateMetadata> {
        self.generate_history.lock().unwrap().clone()
    }

    /// Get consume history.
    pub fn get_consume_history(&self) -> Vec<String> {
        self.consume_history.lock().unwrap().clone()
    }
}

impl StateManager for MockStateManager {
    fn generate(&self, metadata: StateMetadata) -> String {
        self.generate_history.lock().unwrap().push(metadata.clone());

        let state = self
            .next_state
            .lock()
            .unwrap()
            .take()
            .unwrap_or_else(|| format!("mock-state-{}", rand::random::<u32>()));

        self.states.lock().unwrap().insert(state.clone(), metadata);
        state
    }

    fn consume(&self, state: &str) -> Option<StateMetadata> {
        self.consume_history
            .lock()
            .unwrap()
            .push(state.to_string());
        self.states.lock().unwrap().remove(state)
    }

    fn exists(&self, state: &str) -> bool {
        self.states.lock().unwrap().contains_key(state)
    }

    fn clear_expired(&self) {
        // No-op for mock
    }
}

/// Validate state parameter.
pub fn validate_state(
    received: &str,
    state_manager: &dyn StateManager,
) -> Result<StateMetadata, OAuth2Error> {
    state_manager.consume(received).ok_or_else(|| {
        OAuth2Error::Authorization(AuthorizationError::StateMismatch {
            expected: "valid state".to_string(),
            received: received.to_string(),
        })
    })
}

/// Create in-memory state manager.
pub fn create_state_manager() -> impl StateManager {
    InMemoryStateManager::new()
}

/// Create mock state manager for testing.
pub fn create_mock_state_manager() -> MockStateManager {
    MockStateManager::new()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_state_generation() {
        let manager = InMemoryStateManager::new();
        let metadata = StateMetadata::new(
            "https://example.com/callback".to_string(),
            vec!["openid".to_string()],
        );

        let state = manager.generate(metadata);
        assert!(!state.is_empty());
        assert!(manager.exists(&state));
    }

    #[test]
    fn test_state_consumption() {
        let manager = InMemoryStateManager::new();
        let metadata = StateMetadata::new(
            "https://example.com/callback".to_string(),
            vec!["openid".to_string()],
        );

        let state = manager.generate(metadata);
        assert!(manager.exists(&state));

        let consumed = manager.consume(&state);
        assert!(consumed.is_some());
        assert_eq!(consumed.unwrap().redirect_uri, "https://example.com/callback");

        // State should be consumed
        assert!(!manager.exists(&state));
        assert!(manager.consume(&state).is_none());
    }

    #[test]
    fn test_mock_state_manager() {
        let manager = MockStateManager::new();
        manager.set_next_state("test-state-123".to_string());

        let metadata = StateMetadata::new(
            "https://example.com/callback".to_string(),
            Vec::new(),
        );

        let state = manager.generate(metadata);
        assert_eq!(state, "test-state-123");

        let history = manager.get_generate_history();
        assert_eq!(history.len(), 1);
    }
}
